### PR TITLE
Unit tests for the UsernamePasswordAuthenticationCredentials class

### PR DESCRIPTION
**Pull request by vahidR**
Here is the unit tests for the UsernamePasswordAuthenticationCredentials class.
_Junit_ and _Mockito_ libraries has been called and used.

### CODE REVIEW

**Comment by vahidR** [comment]
I removed an unnecessary blank line between two tests and rebased the commits.

**Comment by thc202** [comment]
Reopening as feedback will be provided. Sorry about that.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
+package org.zaproxy.zap.authentication;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.zaproxy.zap.extension.api.ApiResponse;
+import org.zaproxy.zap.extension.api.ApiResponseSet;
+
+import static org.hamcrest.Matchers.*;
```
Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Please use specific imports.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import static org.junit.Assert.*;
+
```
Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Same

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        username = null;
+        password = "something";
+        String parentStringSeparator = "|";
+        usernamePasswordAuthenticationCredentials = new UsernamePasswordAuthenticationCredentials(username, password);
+
+        // When
+        String encodedResult = usernamePasswordAuthenticationCredentials.encode(parentStringSeparator);
+
+        // Then
+        assertThat(encodedResult, is("AA=="));
```
Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Doesn't match function name

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        String parentStringSeparator = "|";
+
+        // When
+        String encodedUsernamePassword = usernamePasswordAuthenticationCredentials.encode(parentStringSeparator);
+
+        // Then
+        assertThat(encodedUsernamePassword, not(isEmptyOrNullString()));
+        assertThat(encodedUsernamePassword, allOf(containsString("dXNlcm5hbWU="),
+                                                    containsString("cGFzc3dvcmQ="),
+                                                    containsString("~")));
```
Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: If hyphen is the "right" Separator it might be good to declare it as such and then reference it here

**Comment by vahidR** [comment]
Thanks guys for your feedbacks. I do appreciate them..
I'll refactor the code to reflect your ideas.

Cheers,

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
+package org.zaproxy.zap.authentication;
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: - The file name should end with UnitTest (to be run by default in Travis builds).
 - It's missing the license header.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    private String username = "username";
+    private String password = "password";
+
+    @Before
+    public void setUp() {
+        this.usernamePasswordAuthenticationCredentials = new UsernamePasswordAuthenticationCredentials(username, password);
+    }
+
+    @Test
+    public void shouldNotUsernameBeEmptyOrNull() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Probably better to rename the test, as is it might give the impression that the `UsernamePasswordAuthenticationCredentials` is created always with a non null/empty username (which might not be the case).

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        // Given
+        String username = usernamePasswordAuthenticationCredentials.getUsername();
+
+        // When/Then
+        assertThat(username, not(isEmptyOrNullString()));
+    }
+
+    @Test
+    public void shouldNotPasswordBeEmptyOrNull() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Same comment for this test.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        // Given
+        String password = usernamePasswordAuthenticationCredentials.getPassword();
+
+        // When/Then
+        assertThat(password, not(isEmptyOrNullString()));
+    }
+
+    @Test
+    public void shouldReturnTureIfIsConfigured() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: There's a typo in "True", but how about "shouldBeConfiguredIfItHasNonNullUsernameAndPassword"? That tells when it is expected to be configured.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+        // When
+        String encodedResult = usernamePasswordAuthenticationCredentials.encode(parentStringSeparator);
+
+        // Then
+        assertThat(encodedResult, is("AA=="));
+    }
+
+    @Test(expected = AssertionError.class)
+    public void shouldParentStringSeparatorNotBeFieldSeparatorWhenEncode() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: This test assumes that the assertions are enabled (most likely they are not).

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                                                    containsString("~")));
+    }
+
+    @Test
+    public void shouldApiResponseRepresentationReturnApiResponseWithValidNameAndJsonFormat() throws Exception {
+        // Given
+        ApiResponse apiResponse = usernamePasswordAuthenticationCredentials.getApiResponseRepresentation();
+
+        // When/Then
+        assertThat(apiResponse.getName(), not(isEmptyOrNullString()));
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Why not assert the actual name ("credentials")?

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        assertThat(apiResponse.toJSON(), notNullValue());
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Could assert that the data returned is correct/expected.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    @Test
+    public void shouldApiResponseRepresentationCallApiResponseSet() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Does it worth verify which `ApiResponse` implementation is being used? (That can be changed while still returning the expected data/format.)

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 8 above)
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Also, good to test when not configured.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 1 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Done..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 3 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Refactored.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 4 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: I'm going to make it broader..
By making a list of non-field-separator samples, randomly picking an element from it, and feeding it to the `encode()` method. 
I will refactor the method name to reflect this

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 5 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Done

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 12 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Probably not.. I may remove this test

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 11 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Refactored..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 10 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Refactored..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 9 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Removed..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 8 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Added

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 6 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Test removed as it was misleading..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 7 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: Test removed as it was misleading..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java (hunk 2 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsTest.java]: done

**Comment by vahidR** [comment]
I tried to address your comments.
Please tell me your idea :)

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        String encodedResult = usernamePasswordAuthenticationCredentials.encode(stringSeparator);
+
+        // Then
+        assertThat(encodedResult, is(nullPattern));
+    }
+
+    @Test
+    public void shouldEncodeUsernameAndPasswordWithANonFieldSeparator() throws Exception {
+        // Given
+        List<String> sampleNonFieldSeparators = Arrays.asList("-", "|", "/", "\\", "+");
```
Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: I'm probably missing something subtle here but in the previous test the pipe character was specifically used, suggesting it's the correct separator. This test seems to suggest it's among the wrong choices....

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        // Given/When
+        ApiResponse apiResponse = usernamePasswordAuthenticationCredentials.getApiResponseRepresentation();
+        JSON jsonRepresentation = apiResponse.toJSON();
+
+        // Then
+        assertThat(apiResponse, notNullValue());
+        assertThat(apiResponse.getName(), equalToIgnoringCase("credentials"));
+        assertThat(jsonRepresentation.toString(), containsString("myUsername"));
+        assertThat(jsonRepresentation.toString(), containsString("myPassword"));
+        assertThat(jsonRepresentation.toString(), containsString("type"));
```
Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Is it worth assigning `jsonRepresentation.toString()` to a String variable instead of converting it 3 times?

Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: You are right.. refactored

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java (hunk 13 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: I don't think so. I think the misunderstanding is because of the bad naming. 
Maybe renaming `sampleNonFieldSeparators` to `someCorrectSeparators` is helpful? 

The method name also can be refactored to `shouldEncodeUsernameAndPasswordWithTheCorrectFieldSeparator()`.

Comment (kingthorin) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Sounds good

Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Alright.. let's rebase and update the PR

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        assertThat(encodedUsernamePassword, notNullValue());
+        assertThat(encodedUsernamePassword, is("bXlVc2VybmFtZQ==~bXlQYXNzd29yZA==~"));
+    }
+
+    @Test
+    public void shouldApiResponseRepresentationReturnApiResponseWithValidNameAndJsonFormat() throws Exception {
+        // Given/When
+        ApiResponse apiResponse = usernamePasswordAuthenticationCredentials.getApiResponseRepresentation();
+        JSON jsonRepresentation = apiResponse.toJSON();
+        String jsonResult= "{\"password\":\"myPassword\",\"type\":\"UsernamePasswordAuthenticationCredentials\",\"username\":\"myUsername\"}";
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: I'm getting an error in this test:
```
Expected: is "{\"password\":\"myPassword\",\"type\":\"UsernamePasswordAuthenticationCredentials\",\"username\":\"myUsername\"}"
     but: was "{\"username\":\"myUsername\",\"type\":\"UsernamePasswordAuthenticationCredentials\",\"password\":\"myPassword\"}"
```

Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Weird.. I'm getting the exact order and my tests pass. 
Maybe it's better to search for fields with `allOf()` matcher and check for the specific values..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+        // Then
+        assertThat(encodedResult, is(nullPattern));
+    }
+
+    @Test
+    public void shouldEncodeUsernameAndPasswordWithTheCorrectFieldSeparator() throws Exception {
+        // Given
+        List<String> someCorrectSeparators = Arrays.asList("-", "|", "/", "\\", "+");
+        String aCorrectSeparator = someCorrectSeparators.get(new Random().nextInt(someCorrectSeparators.size()));
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: It could test all separators, something like:
```Java
for (String correctSeparator : someCorrectSeparators) {
    // When
    String encodedUsernamePassword = usernamePasswordAuthenticationCredentials.encode(correctSeparator);

    // Then
    assertThat(String.format("Failed to encode with '%s'", correctSeparator), encodedUsernamePassword, notNullValue());
    assertThat(
            String.format("Failed to properly encode with '%s'", correctSeparator),
            encodedUsernamePassword,
            is("bXlVc2VybmFtZQ==~bXlQYXNzd29yZA==~"));
}
```
(In any case it should indicate which separator was used.)

Also, maybe we should throw an exception instead of silently continue the encoding when "wrong" parent separator is used?

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java (hunk 15 above)
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Might have to do with the locale?
Yeah, that seems more robust.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java (hunk 16 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: > It could test all separators, something like:

 That is another option. I'll change the test to reflect this.

> Also, maybe we should throw an exception instead of silently continue the encoding when "wrong" parent separator is used?

In the class we have an assert statement : `assert (!FIELD_SEPARATOR.equals(parentStringSeparator));`
I removed the unit test for it as the assertions are most likely not enabled).
I will refactor the code to address this Exception and write a unit test for it ..

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java (hunk 15 above)
Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Maybe.. I'll go with the `allOf()` approach

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java (hunk 16 above)
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: OK, thanks.

Comment (vahidR) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: A quick note: In the code example that you proposed, we are testing the **Correctness** of the separators. That's why we have `someCorrectSeparators` to emphasis the correctness.

**Comment by vahidR** [comment]
When I debug the `throw new IllegalArgumentException(Constant.messages.getString("...");` within the `UsernamePasswordAuthenticationCredentials` class, I get this error : 

`Constant.messages -> Class not loaded : org.parosproxy.paros.Constant`

Any idea why it can't load the class ?

**Comment by vahidR** [comment]
@thc202 @kingthorin any idea about the above comment ?

**Comment by kingthorin** [comment]
Not off the top of my head, however I can say that Exception messages aren't expected to be internationalized. So you can just provide/return the string directly when throwing. e.g:
` throw new IllegalArgumentException("The string separator must not be the same as Field Separator (~)");`

**Comment by thc202** [comment]
How are you running the tests?

Yeah, that exception should not be internationalised (if that happens it's a bug in the code, better not have to translate that).

**Comment by vahidR** [comment]
@kingthorin Nice... will do that. Thanks for info :)

**Comment by vahidR** [comment]
All Done ..

**Comment by vahidR** [comment]
> How are you running the tests?

That specific test failed on my local machine, and as it was about loading the `paros`, I tried to push/load it on the CI environment to verify my problem. 
Sorry for the inconvenience..

**Comment by kingthorin** [comment]
Looks good to me

File: src/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentials.java [hunk 17]
```diff
@@ -84,7 +84,9 @@ public boolean isConfigured() {
 
 	@Override
 	public String encode(String parentStringSeparator) {
-		assert (!FIELD_SEPARATOR.equals(parentStringSeparator));
+		if (FIELD_SEPARATOR.equals(parentStringSeparator)) {
+			throw new IllegalArgumentException("The string separator must not be the same as Field Separator (~).");
```
Comment (thc202) [inline src/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentials.java]: Better to append/use FIELD_SEPARATOR.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    private String password = "myPass";
+
+    @Before
+    public void setUp() {
+        this.usernamePasswordAuthenticationCredentials = new UsernamePasswordAuthenticationCredentials(username, password);
+        this.notConfiguredInstance = new UsernamePasswordAuthenticationCredentials();
+    }
+
+    @Test
+    public void shouldBeConfiguredIfUsernameAndPasswordAreNotNull() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: The "throws Exception" should be removed, not necessary.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        // Given/When
+        boolean isConfigured = usernamePasswordAuthenticationCredentials.isConfigured();
+
+        // Then
+        assertThat(isConfigured, is(true));
+    }
+
+    @Test
+    public void shouldNotBeConfiguredIfUsernameAndPasswordAreNull() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: Be good to also test with (just) the password null.

File: test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    @Test
+    public void shouldApiResponseRepresentationReturnApiResponseWithValidNameAndJsonFormat() throws Exception {
+        // Given/When
+        ApiResponse apiResponse = usernamePasswordAuthenticationCredentials.getApiResponseRepresentation();
+        JSON jsonRepresentation = apiResponse.toJSON();
+
+        // Then
+        assertThat(apiResponse, notNullValue());
+        assertThat(apiResponse.getName(), equalToIgnoringCase("credentials"));
+        assertThat(jsonRepresentation.toString(), allOf(containsString("myPass"), containsString("myUser"), containsString("type")));
```
Comment (thc202) [inline test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java]: - Should check "UsernamePasswordAuthenticationCredentials" instead of "type"? (or also add the former);
 - The "myPass"  and "myUser" should be extracted to a constant (and the corresponding variables initialised with those), less changes if one needs to be changed.

**Comment by kingthorin** [comment]
@vahidr are you able to address @thc202's feedback?

**Comment by vahidR** [comment]
@kingthorin I start to work on it in 10 days.. 
Sorry, I've been overloaded with multiple tasks during past weeks, and couldn't invest enough time on it..

**Comment by kingthorin** [comment]
Ya no worries, let us know if there's something we can do to help.

**Comment by kingthorin** [comment]
@vahidr are you able to address @thc202's feedback?

**Comment by vahidR** [comment]
Duh... totally missed this thread. I will apply the comments..

**Comment by vahidR** [comment]
Unfortunately, I'm going to close this thread, as I didn't invest enough time on it. 
I hope I can get back soon :)

**Comment by thc202** [comment]
OK, although, if you do not have changes in progress locally, would you mind allow us to do the final tweaks to merge it?

**Comment by vahidR** [comment]
@thc202 Sure, go ahead. The important thing is the software itself..

**Comment by thc202** [comment]
Great, thank you. Can you confirm that we have write permission to this pull request's branch? (I don't see any indication about that...)

**Comment by vahidR** [comment]
I just did... Feel free to change it :)

**Comment by thc202** [comment]
Thanks, will do :)

**Comment by thc202** [comment]
Rebased the branch and addressed the previous comments. Following the diff of the changes:
```diff
diff --git a/src/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentials.java b/src/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentials.java
index 3706e8b..9287de1 100644
--- a/src/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentials.java
+++ b/src/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentials.java
@@ -85,7 +85,8 @@ class UsernamePasswordAuthenticationCredentials implements AuthenticationCredent
 	@Override
 	public String encode(String parentStringSeparator) {
 		if (FIELD_SEPARATOR.equals(parentStringSeparator)) {
-			throw new IllegalArgumentException("The string separator must not be the same as Field Separator (~).");
+			throw new IllegalArgumentException(
+					"The string separator must not be the same as Field Separator (" + FIELD_SEPARATOR + ").");
 		}
 		if (username == null) {
 			return NULL_CREDENTIALS;
diff --git a/test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java b/test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java
index 02f6e0f..d2076b3 100644
--- a/test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java
+++ b/test/org/zaproxy/zap/authentication/UsernamePasswordAuthenticationCredentialsUnitTest.java
@@ -51,7 +51,7 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
     }
 
     @Test
-    public void shouldBeConfiguredIfUsernameAndPasswordAreNotNull() throws Exception {
+    public void shouldBeConfiguredIfUsernameAndPasswordAreNotNull() {
         // Given/When
         boolean isConfigured = usernamePasswordAuthenticationCredentials.isConfigured();
 
@@ -60,7 +60,7 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
     }
 
     @Test
-    public void shouldNotBeConfiguredIfUsernameAndPasswordAreNull() throws Exception {
+    public void shouldNotBeConfiguredIfUsernameAndPasswordAreNull() {
         // Given/When
         boolean isConfigured = notConfiguredInstance.isConfigured();
 
@@ -68,8 +68,18 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
         assertThat(isConfigured, is(false));
     }
 
+    @Test
+    public void shouldNotBeConfiguredIfPasswordIsNull() {
+        // Given
+        UsernamePasswordAuthenticationCredentials credentials = new UsernamePasswordAuthenticationCredentials(username, null);
+        // When
+        boolean isConfigured = notConfiguredInstance.isConfigured();
+        // Then
+        assertThat(isConfigured, is(false));
+    }
+
     @Test(expected = IllegalArgumentException.class)
-    public void shouldThrowExceptionWhileEncodeWithFieldSeparator() throws Exception {
+    public void shouldThrowExceptionWhileEncodeWithFieldSeparator() {
         // Given
         String fieldSeparator = "~";
         usernamePasswordAuthenticationCredentials = new UsernamePasswordAuthenticationCredentials(username, password);
@@ -81,7 +91,7 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
     }
 
     @Test
-    public void shouldEncodeMethodReturnNullPatternIfUsernameIsNull() throws Exception {
+    public void shouldEncodeMethodReturnNullPatternIfUsernameIsNull() {
         // Given
         String nullPattern = "AA==";
         username = null;
@@ -97,7 +107,7 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
     }
 
     @Test
-    public void shouldEncodeUsernameAndPasswordWithTheCorrectFieldSeparator() throws Exception {
+    public void shouldEncodeUsernameAndPasswordWithTheCorrectFieldSeparator() {
         // Given
         List<String> someCorrectSeparators = Arrays.asList("-", "|", "/", "\\", "+");
 
@@ -114,7 +124,7 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
     }
 
     @Test
-    public void shouldApiResponseRepresentationReturnApiResponseWithValidNameAndJsonFormat() throws Exception {
+    public void shouldApiResponseRepresentationReturnApiResponseWithValidNameAndJsonFormat() {
         // Given/When
         ApiResponse apiResponse = usernamePasswordAuthenticationCredentials.getApiResponseRepresentation();
         JSON jsonRepresentation = apiResponse.toJSON();
@@ -122,7 +132,13 @@ public class UsernamePasswordAuthenticationCredentialsUnitTest {
         // Then
         assertThat(apiResponse, notNullValue());
         assertThat(apiResponse.getName(), equalToIgnoringCase("credentials"));
-        assertThat(jsonRepresentation.toString(), allOf(containsString("myPass"), containsString("myUser"), containsString("type")));
+        assertThat(jsonRepresentation.toString(), allOf(
+                containsString("username"),
+                containsString(username),
+                containsString("password"),
+                containsString(password),
+                containsString("type"),
+                containsString("UsernamePasswordAuthenticationCredentials")));
 
     }
 }
\ No newline at end of file
```
